src/fundamental_interface.jl

```diff
@@ -0,0 +1,275 @@
+
+julia> parent(x) == S
+true
+```
+"""
+function parent end
+
+# TODO: Give example
+@doc Markdown.doc"""
+    elem_type(parent)
+    elem_type(parent_type)
+
+Given a parent object (or its type), return the type of its elements.
+
+# Example
+```jldoctest; setup = :(using AbstractAlgebra)
+julia> S, x = PowerSeriesRing(QQ, 2, "x")
+(Univariate power series ring in x over Rationals, x + O(x^3))
+
+julia> elem_type(S) == typeof(x)
+true
+```
+"""
+elem_type(x)  = elem_type(typeof(x))
+elem_type(T::DataType) = throw(MethodError(elem_type, (T,)))
+
+@doc Markdown.doc"""
+    parent_type(element)
+    parent_type(element_type)
+
+Given an element (or its type), return the type of its parent object.
+
+# Example
+```jldoctest; setup = :(using AbstractAlgebra)
+julia> R, x = PolynomialRing(ZZ, "x")
+(Univariate Polynomial Ring in x over Integers, x)
+
+julia> S = MatrixSpace(R, 2, 2)
+Matrix Space of 2 rows and 2 columns over Univariate Polynomial Ring in x over Integers
+
+julia> a = rand(S, 0:1, 0:1);
+
+julia> parent_type(a) == typeof(S)
+true
+```
+"""
+parent_type(x) = parent_type(typeof(x))
+parent_type(T::DataType) = throw(MethodError(parent_type, (T,)))
+
+@doc Markdown.doc"""
+    base_ring(a::T)
+
+Return base ring $R$ of given element or parent $a$.
+
+# Examples
+```jldoctest; setup = :(using AbstractAlgebra)
+julia> S, x = PolynomialRing(QQ, "x")
+(Univariate Polynomial Ring in x over Rationals, x)
+
+julia> base_ring(S) == QQ
+true
+
+julia> R = GF(7)
+Finite field F_7
+
+julia> base_ring(R)
+Union{}
+```
+"""
+function base_ring end
+
+###############################################################################
+#
+#   Special elements
+#
+###############################################################################
+
+@doc Markdown.doc"""
+    one(a::T)
+
+Return the (multiplicative) identity in the family of $a$.
+
+# Examples
+```jldoctest; setup = :(using AbstractAlgebra)
+julia> S = MatrixSpace(ZZ, 2, 2)
+Matrix Space of 2 rows and 2 columns over Integers
+
+julia> one(S)
+[1   0]
+[0   1]
+
+julia> R, x = PuiseuxSeriesField(QQ, 4, "x")
+(Puiseux series field in x over Rationals, x + O(x^5))
+
+julia> one(x)
+1 + O(x^4)
+
+julia> G = GF(5)
+Finite field F_5
+
+julia> one(G)
+1
+```
+"""
+function one end
+
+@doc Markdown.doc"""
+    zero(a::T)
+
+Return the zero or additive identity in the family of $a$.
+
+# Examples
+```jldoctest; setup = :(using AbstractAlgebra)
+julia> S = MatrixAlgebra(QQ, 2)
+Matrix Algebra of degree 2 over Rationals
+
+julia> zero(S)
+[0//1   0//1]
+[0//1   0//1]
+
+julia> R, x = PolynomialRing(ZZ, "x")
+(Univariate Polynomial Ring in x over Integers, x)
+
+julia> zero(x^3 + 2)
+0
+```
+"""
+function zero end
+
+###############################################################################
+#
+#   Basic manipulation
+#
+###############################################################################
+
+@doc Markdown.doc"""
+    isone(a::T)
+
+Return true if $a$ is an identity, else return false.
```

Comment:
At the very least, need to do:
```suggestion
Return true if $a$ is a multiplicative identity, else return false.
```
But it probably also needs to refer to "its family" (or rather: it's parent?)

Reply:
My vision (and I guess yours too) is that AbstractAlgebra will be a building stone for many pure mathematics applications in Julia.

Since one can have identity maps and so on, which are not multiplicative identities but rather just identities, I thought this was a better fit. But at the moment, this has no application, so we could go with the suggested change and later change it if it fits the purpose.

Reply:
Perhaps that kind of identity then should better be called "identity", as `one` has strong association with multiplicative notation. Also, how would a "generic identity without a prescribed corresponding operation" even be used in generic code? (As for non-generic code that only deals with one particular algebraic structure, it doesn't need to be covered by generic documentation, does it?)

Based on experience, I am also quite wary about attempting generalizations without immediate applications. 

That said, I really mean these as open questions; while I have strong believes on these, I am happy to have my believes challenged and overturned, they are just believes after all, not absolute truths :-). So please don't be discouraged by my prying questions

Reply:
No, I think discussions (even heated ones) are good ;)

I like your opinion on `one` vs `identity`. I think we should go that way. So should additive group only have `zero` defined, only `one` defined, or do they have both defined to be the same?

Reply:
Sorry, didn't know that `AdditiveGroup` existed in AA. Will fix accordingly.